### PR TITLE
Renewal Playground with WebContainer API

### LINKED ISSUE #734: Renewal Playground
Continued from #214

Renewal playground website.
https://playground.markuplint.dev/

## Planning features

- [ ] Merge it into the main website (markuplint.dev).
  - in order to make it easier to navigate around the website.
- [ ] Show the rules of the currently applied configuration. 
- [ ] Allow users to select rules with GUI.
- [ ] Allow users to resize each pane.
- [ ] Show the errors and the warnings.
- [ ] Add some examples (e.g. BEM pattern class-naming config)

## Screenshot of the draft PR by @YusukeHirao 

スクリーンショット_20230223_002408

### PR DESCRIPTION
**Pull request by kagankan**
Fixes #734

## What is the purpose of this PR?

- [ ] Fix bug
- [ ] Fix typo
- [ ] Update specs
- [ ] Add new rule
- [ ] Add new parser
- [ ] Improve or refactor rules
- [ ] Add a new core feature
- [ ] Improve or refactor core features
- [ ] Update documents
- [x] Others

### Description

Brand New Playground!

- Use React with Vite
- Use WebContainer API
  - [Stylelint Demo](https://stylelint.io/demo/) was used as a reference.
  - The WebContainer API runs a Node server in the browser. Install the `markuplint` package in the browser.
- Enable to change the file type of code.
- Enable to change the config.
- Enable to list errors in the Problems tab.
- Add some examples 
- Use TailwindCSS
- Allow users to select rules with GUI.

image

### CODE REVIEW

**Comment by kagankan** [comment]
I don't know why the Playground Building Test fails... 🤔 
`yarn pg:build` works fine on my local Windows.

**Comment by YusukeHirao** [comment]
@kagankan Thanks for the PR. 

The reason the tests are failing is because the workspace was removed. In the Actions, only the workspace installation is being performed, so the package.json in the Playground directory is not being checked, and no installation is happening there.

**Comment by kagankan** [comment]
@YusukeHirao 

Oh, I misunderstood `pg:build` in the root and `pg:build` in the playground.

The test was removed in the dev branch, so I have pulled in that change.
Please review the code. 🙏

File: playground/package.json [hunk 1]
```diff
@@ -1,59 +1,46 @@
...
-		"@markuplint/rules": "3.10.0",
-		"monaco-editor": "0.38.0",
-		"svelte": "3.58.0",
-		"tslib": "2.5.0"
+		"@headlessui/react": "^1.7.14",
+		"@markuplint/parser-utils": "^3.7.0",
+		"@monaco-editor/react": "^4.5.0",
+		"@webcontainer/api": "^1.1.3",
+		"ansi-regex": "^6.0.1",
+		"markuplint": "^3.8.0",
```
Comment (YusukeHirao) [inline playground/package.json:19]: @kagankan 

Could you please explain the relationship between the packages needed here and those referenced in `deps.json`?

https://github.com/markuplint/markuplint/pull/973/files#diff-63fccff55aced03b2348fd7c4970b2d86a885b8ed64cec1cd81c686ae7e0d7d2

Comment (kagankan) [reply playground/package.json:19]: @YusukeHirao 


`playground/package.json` defines packages installed during development and build.
On the other hand, `deps.json` defines packages installed on the browser (via WebContainer API). The `markuplint` defined in this `deps.json` is executed as a linter on the playground site.

Only type information (`MLResultInfo`) is used from `markuplint` installed in `playground/package.json`.
https://github.com/markuplint/markuplint/pull/973/files#diff-c34e3203108664d9deef6eb19163ebbc58121db73b79dc4831784369bf3dcc03

So, if there is a change in the type information, a runtime error may occur, but I'm assuming no change in that.
(Please let me know if there is an appropriate way to handle this)

Comment (kagankan) [reply playground/package.json:19]: @YusukeHirao 

Do you have any additional questions?
Please let me know if my explanation is not clear enough.

**Comment by YusukeHirao** [comment]
@kagankan 

I apologize for the inconvenient timing. I've changed the dev branch for v4, including transitioning to ESM. I'd appreciate it if you could continue with the implementation in line with these changes. Please let me know if you have any questions.

**Comment by YusukeHirao** [comment]
@kagankan 

FYI:

It fails to load in the ESM system because `@markuplint/html-spec` is a CommonJS package.

It may resolve if it loads the JSON directly.

https://github.com/markuplint/markuplint/blob/f34ad833e66ef2e27a0ac77509e2d132eed8dbd4/playground/vite.config.js#L2-L6

**Comment by kagankan** [comment]
コメントいただいてから放置しててすみません、これからv4に向けた修正検討しますね

**Comment by kagankan** [comment]
@YusukeHirao 
Not only considering v4, but also brushing up the functions takes time, sorry. 🙏 

For v4 on Playground, I'm thinking as follows:

- Since WebContainer API is used, `npm install` is executed on the playground site. Therefore, it does not depend on the `packages/**` directories.
- Make it possible to choose `latest` or `next` on the playground.
  - This allows users to try both versions.

Is it okay?
Please let me know if you have any concerns.

Here's the progress:
(WIP. I will improve the UI and functions more)
https://655f3f87ca55f60008e17941--aquamarine-croquembouche-850314.netlify.app/

Video (changing the version from `latest` to `next`)
https://github.com/markuplint/markuplint/assets/34136752/8a0ae274-c4ef-4c4c-a269-23d6090c1c82

**Comment by kagankan** [comment]
@YusukeHirao 
I have updated the PR. Please review it again!
(Sorry for the many diff files 🙏 )

Demo:
https://65759ef4d209700008d18603--aquamarine-croquembouche-850314.netlify.app/

**Comment by YusukeHirao** [comment]
@kagankan 

I checked it. LGTM;)

Do you have any other remaining tasks?
I want it to highlight codes besides HTML.

**Comment by kagankan** [comment]
@YusukeHirao 

Thanks!!!

I pushed the change for the URL hash shortening (closes #1290).
That's all I want to do in this PR.

>I want it to highlight codes besides HTML.

I want to do this too...!
I tried a little, but it seems that it is not so easy to do. 😢 
(To be exact, I could highlight JSX code, but the color theme was unnatural...)
Please let me fix it in another branch in the future.

**Comment by YusukeHirao** [comment]
@kagankan 

> Please let me fix it in another branch in the future.

OK, Thank you!

**Review by YusukeHirao — APPROVED** [review]

**Comment by YusukeHirao** [comment]
@kagankan 

Could you tell me what I should specify to the Netlify deploy setting?

**Comment by kagankan** [comment]
@YusukeHirao 

Please set up as follows.

Base directory: `playground`
Package directory: don't need to set (same as base directory)
Build command: `yarn run pg:build`
Publish directory: `playground/dist`
Functions directory: don't need to set ({base directory}/netlify/functions)
Deploy log visibility: Either is OK
Build status: `Active builds`

Here is a screenshot of my Netlify build settings.

image

**Pull request merged**